delete_old_unclaimed_attachments: Improve performance & error handling.

abdelrahman725 opened the pull request:
Fixes #30869

[CZO discussion](https://chat.zulip.org/#narrow/channel/3-backend/topic/smarter.20do_delete_old_unclaimed_attachments)

- bulk update and delete old unclaimed attachments, improving performance.

- smarter algorithm for deleting attachments, to avoid potential leaking attachments in db or storage:

  1- mark the attachment objects as "deleted".
  2- bulk-delete them from storage.
  3- delete the actual Attachment objects from the database.

alya commented:
Thanks! Please complete the self-review checklist (as applicable), and post a comment to request a review when ready.

timabbott commented on zerver/models/messages.py:
```diff
@@ -670,6 +670,9 @@ class ImageAttachment(models.Model):
     frames = models.IntegerField()
 
+    # indicates whether we attempted to delete this image
+    deleted = models.BooleanField(default=False)
```
These should probably have `db_index=True`, so it's easy to query them without a table scan.

timabbott replied:
Why did you not do `db_index=True` here, and instead add a custom index?

abdelrahman725 replied:
Please read my [reply](https://github.com/zulip/zulip/pull/32563#discussion_r1871284319) about using `partial index` and its rationale, but in short :
 
`db_index=True` is not used when creating a partial index

timabbott commented on zerver/actions/uploads.py:
```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
+    """delete old unclaimed attachments from database."""
+    Attachment.objects.filter(deleted=True).delete()
+    ArchivedAttachment.objects.filter(deleted=True).delete()
+    ImageAttachment.objects.filter(deleted=True).delete()
```
See below, but these queries are inefficient without an index. You can generally use `.explain()` (printing it out) to check query plans in a management shell.

abdelrahman725 replied:
**Testing after implementing partial index on `deleted=True`  and having `460` attachments** : 

We can see the index is now used when filtering by `deleted=True`, and the total estimated cost in the case of index `12` is less than that without index `24` 
 
```python
Attachment.objects.filter(deleted=False).explain() 
'Seq Scan on zerver_attachment  (cost=0.00..24.59 rows=457 width=122)\n  Filter: (NOT deleted)'

Attachment.objects.filter(deleted=True).explain()
'Index Scan using zerver_attachment_deleted_true_idx on zerver_attachment  (cost=0.13..12.32 rows=2 width=122)'

```

timabbott commented on zerver/actions/uploads.py:
```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
-
-    if storage_paths:
-        delete_message_attachments(storage_paths)
+
```
The logic above likely should be querying everything with `deleted=True`, not the set of things passed in, if we want failures to be handled properly, right?

abdelrahman725 replied:
You mean instead of passing `storage_paths` we should pass `path_id`s of all `deleted=True` objects? 

But there are some `path_id`s that don't have a corresponding db object, like [thumbnail path](https://github.com/zulip/zulip/blob/main/zerver/actions/uploads.py#L98-L99)

timabbott commented on zerver/actions/uploads.py:
```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
+    for batch_start_index in range(0, len(storage_paths), DELETE_BATCH_SIZE):
+        delete_message_attachments(
+            storage_paths[batch_start_index : batch_start_index + DELETE_BATCH_SIZE]
+        )
```
Does `delete_message_attachments` continue properly in both backends if some of the paths don't exist?

abdelrahman725 replied:
good question !
## for local backend
it continues yes, and just raises this warning
```python
logging.warning("%s does not exist. Its entry in the database will be removed.", file_name)
```
The way I tested this is by making an early deletion for a file using `delete_local_file` before calling `delete_message_attachments`
## for S3 backend

I think yes, function that's used is [S3UploadBackend.delete_message_attachments](https://github.com/zulip/zulip/blob/main/zerver/lib/upload/s3.py#L286-L289)

but there are no warnings for non-existing files, because [Note that if the object specified in the request is not found, Amazon S3 returns the result as deleted.](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3/bucket/delete_objects.html#:~:text=Note%20that%20if%20the%20object%20specified%20in%20the%20request%20is%20not%20found%2C%20Amazon%20S3%20returns%20the%20result%20as%20deleted.)

abdelrahman725 commented:
## Updates : 
Added partial index `deleted=True` to `ArchivedAttachment`, `Attachment` and `ImageAttachment`

abdelrahman725 commented:
## Updates : 
call `get_old_unclaimed_attachments` only once (instead of twice) inside `Command.hanlde()`, which is more efficient as `get_old_unclaimed_attachments` executes 2 fairly  complex queries

abdelrahman725 commented on zerver/tests/test_delete_unclaimed_attachments.py:
```diff
@@ -55,6 +56,10 @@ def assert_exists(
             ArchivedAttachment.objects.filter(id=attachment.id).exists(), has_archived_attachment
         )
 
+    def get_and_do_delete_old_unclaimed_attachments(self, weeks_ago: int) -> None:
+        old_attachments, old_archived_attachments = get_old_unclaimed_attachments(weeks_ago)
+        do_delete_old_unclaimed_attachments(old_attachments, old_archived_attachments)
+
```
This new function is only introduced to achieve consistency and to avoid having to write `get_old_unclaimed_attachments` each time we call  `do_delete_old_unclaimed_attachments`  in each of the test cases.

timabbott commented on zerver/actions/uploads.py:
```diff
@@ -80,45 +83,60 @@ def do_claim_attachments(
+    Attachment.objects.filter(deleted=True).delete()
+    ArchivedAttachment.objects.filter(deleted=True).delete()
+    ImageAttachment.objects.filter(deleted=True).delete()
 
```
I'm not sure `ImageAttachment` requires a `deleted` field; shouldn't you be able to just do `ImageAttachment.objects.filter(path_id__in=deleted_attachments_query).delete()`?

abdelrahman725 replied:
Yes, you are right

timabbott replied:
Can you show the precise query plans from `EXPLAIN ANALYZE`?

I expect it to be well-indexed regardless.

abdelrahman725 replied:
yes, `path_id` field, which we would filter by in this case, of `ImageAttachment` is indexed.

running : 
```python3
    query = ImageAttachment.objects.filter(
        Q(path_id__in=attachments_path_ids) | Q(path_id__in=archived_attachments_path_ids)
    ).explain(analyze=True)
```
outputs  in 2 tests: 

```shell
Index Scan using zerver_imageattachment_path_id_0fbe468e_like on zerver_imageattachment  (cost=0.15..8.17 rows=1 width=88) (actual time=0.004..0.004 rows=0 loops=1)
  Index Cond: (path_id = '2/96/zlOvST53OIUJ56oOyxb8oQwn/text.txt'::text)
Planning Time: 0.036 ms
Execution Time: 0.012 ms
```

and output in 1 test :

```shell
Seq Scan on zerver_imageattachment  (cost=0.03..29.28 rows=20 width=88) (actual time=0.006..0.006 rows=0 loops=1)......

Planning Time: 0.051 ms
Execution Time: 0.014 ms
```

The difference in scan type (Seq or index) is normal, because it's up to postgres engine to decide which one is better for each query.

abdelrahman725 committed:
delete_old_unclaimed_attachments: Improve performance & error handling.

Fixes #30869

- bulk update and delete old unclaimed attachments, improving performance.

- call "get_old_unclaimed_attachments()" only once instead of twice
inside "Command.handle()"

- smarter algorithm for deleting attachments:

1- mark the attachment objects as "deleted".
2- bulk-delete them from storage.
3- delete the actual Attachment objects from the database.
